great_expectations/zep/bi_directional_dict.py

```diff
@@ -0,0 +1,62 @@
+import pprint
+from collections import UserDict
+from typing import Hashable, Iterable, Mapping, Optional, TypeVar
+
+T = TypeVar("T", bound=Hashable)
+
+
+class BiDict(
+    UserDict,
+    Mapping[T, T],
+):
+    """
+    Dict-like Mapping object that creates keys from values and values from keys.
+    Because of this, all values must be Hashable.
+
+    If a Mapping-like object is passed as the first parameter, its key/values will be
+    unpacked (and combined with kwargs) into the new `BiDict` object.
+    """
+
+    def __init__(
+        self,
+        __dict: Optional[Mapping[Hashable, Hashable]] = None,
+        **kwargs: Hashable,
+    ):
+        if isinstance(__dict, Mapping):
+            kwargs: dict = {**__dict, **kwargs}  # type: ignore[no-redef]
+            __dict = None
+        super().__init__(__dict, **kwargs)
+
+    def __delitem__(self, key: Hashable):
+        value = self.data.pop(key)
+        super().pop(value, None)
+
+    def __setitem__(self, key: Hashable, value: Hashable):
+        if key in self:
+            del self[self[key]]
+        if value in self:
+            del self[value]
+        super().__setitem__(key, value)
+        super().__setitem__(value, key)
```

Comment:
I initially intended to have this structure be purpose-built for our type-lookout (in which case it would not have allowed this) but then I went and tried to generalize it 😬 .
But speaking to your comment about `register_factory()`, I think it's better if this structure could be responsible for preventing naming conflicts simply by making overwrites impossible.